pkg/operator/vspherecontroller/vspherecontroller.go

```diff
@@ -658,46 +669,61 @@ func (c *VSphereController) createCSISecret(
 
 func (c *VSphereController) applyClusterCSIDriverChange(
 	infra *ocpv1.Infrastructure,
-	sourceCFG vsphere.VSphereConfig,
+	sourceCFG *vclib.VSphereConfig,
 	clusterCSIDriver *operatorapi.ClusterCSIDriver,
 	datastoreURL string) (*corev1.Secret, error) {
 
 	csiConfigString := string(c.csiConfigManifest)
 
-	dataCenterNames, err := utils.GetDatacenters(&sourceCFG)
+	// Validate config.  We used to fail when calling utils.GetDataCenters, but now we will let config validate itself.
+	err := sourceCFG.ValidateConfig(c.featureGates)
+	if err != nil {
+		return nil, err
+	}
+
+	csiVCenterConfigBytes, err := assets.ReadFile("csi_cloud_config_vcenters.ini")
 
 	if err != nil {
 		return nil, err
 	}
 
-	datacenters := strings.Join(dataCenterNames, ",")
+	// Generate cluster id and append all vcenters.  Also need to inject user/pass for vcenters since driver does
+	// not support loading from secret.  It expects user/pass either in the ini file or as an ENV variable.  ENV
+	// variable was used in older, single vcenter way where passed into container from operator.
+	config := sourceCFG.Config
+	var vcenters string
+	for _, vcenter := range config.VirtualCenter {
+		vcenterStr := string(csiVCenterConfigBytes)
+
+		user, password, err := getUserAndPassword(defaultNamespace, cloudCredSecretName, vcenter.VCenterIP, infra, c.configMapLister, c.apiClients.SecretInformer, c.featureGates)
+		if err != nil {
+			return nil, err
+		}
+
+		for pattern, value := range map[string]string{
+			"${VCENTER}":                 vcenter.VCenterIP,
+			"${DATACENTERS}":             vcenter.Datacenters,
+			"${MIGRATION_DATASTORE_URL}": datastoreURL,
+			"${PASSWORD}":                password,
+			"${USER}":                    user,
+		} {
+			vcenterStr = strings.ReplaceAll(vcenterStr, pattern, value)
+		}
+		vcenters = vcenters + "\n" + vcenterStr
+	}
+
 	for pattern, value := range map[string]string{
```

Comment:
So, in general we want secret's content to be "stable". Because - we do not want operator/controllers to be restarted randomly if order of vcenter names in config file changes. If I am not wrong - iterating a golang's map is essentially random and there is no guarantee order in which it will appear and as a result, I think secret can contain vcenters in any random order and this will cause unnecessary restart of controller.

Reply:
I made the changes.  let me know your thoughts.